### PR TITLE
:bug: Fix button not being inject

### PR DESCRIPTION
**Pull request by 1natsu172**
Issue: #31 
Button was not injected in the video with creator-heart-action in the comment.

Compared to other video pages this video type is slow to load comments.
Moreover, when navigating from this video type to another video, it turned out that the rewriting of the parse target node was delayed.

### CODE REVIEW

File: script.js [hunk 1]
```diff
@@ -148,7 +148,15 @@ const newYouTube = {
 
   registerListeners() {
     document.addEventListener('yt-visibility-refresh', newYouTube.inject); // Inject the button on info panel render.
-    document.addEventListener('yt-page-data-updated', newYouTube.injectCommentsCount); // Asynchronously get comments count when navigated to video page
+
+    // Asynchronously get comments count when navigated to video page.
+    document.addEventListener('yt-page-data-updated', function(e) {
+      // setTimeout mean: This event sometimes precedes rewriting of number of comment node. So wait few ms.
+      setTimeout(() => {
```
Comment (1natsu172) [inline script.js]: Case of issue: heart-action typed video watched then next another video.
The next video's parse target comments node is rewritten late.

File: script.js [hunk 2]
```diff
@@ -284,7 +286,14 @@ const newYouTube = {
     const countString = extractDigitArray.join();
     newYouTube._commentsState.hasGotCount = true;
     newYouTube._commentsState.currentCount = countString;
-    newYouTube._rewriteCommentsCount()
+
```
Comment (1natsu172) [inline script.js]: Fetch of the number of comments may end before the button's injection. So wait for the button.

File: script.js [hunk 3]
```diff
@@ -148,7 +148,15 @@ const newYouTube = {
...
-    document.addEventListener('yt-page-data-updated', newYouTube.injectCommentsCount); // Asynchronously get comments count when navigated to video page
+
+    // Asynchronously get comments count when navigated to video page.
+    document.addEventListener('yt-page-data-updated', function(e) {
+      // setTimeout mean: This event sometimes precedes rewriting of number of comment node. So wait few ms.
+      setTimeout(() => {
+        newYouTube.injectCommentsCount(e)
+      }, 500)
+    });
+
```
Comment (1natsu172) [inline script.js:159]: Case of issue: heart-action typed video watched then next another video.
The next video's parse target comments node is rewritten late.

File: script.js [hunk 4]
```diff
@@ -284,7 +286,14 @@ const newYouTube = {
...
     newYouTube._commentsState.currentCount = countString;
-    newYouTube._rewriteCommentsCount()
+    
+    const tryRewrite = setInterval(() => {
+      if (newYouTube._rewriteCommentsCount() === false) {
+        newYouTube._rewriteCommentsCount();
+      } else {
+        clearInterval(tryRewrite);
+      }
+    }, 100);
```
Comment (1natsu172) [inline script.js:298]: Fetch of the number of comments may end before the button's injection. So wait for the button.

**Comment by dideler** [comment]
Checked out the branch and ran it locally but I'm still seeing the issue on [this video](https://www.youtube.com/watch?v=7a9lsGtVziM) of the button not being injecting. I can try having a better look tomorrow night.

**Comment by 1natsu172** [comment]
<s>I thought "really?" And watched it on the UK version of YouTube but it certainly doesn't work... 😨 
It's working fine on YouTube in Japanese version…
I will investigate a little more.</s>

No, this was my mistake, I was in the master branch locally (Master doesn't contain this fix).
I noticed it and tried with this(#31) branch, so it still works on the UK version of YouTube ...

**Comment by dideler** [comment]
Here are the steps I'm using to reproduce the bug:

1. Check out branch `fix#31/support-heart-action-video` with commit 11e7a33
1. Reload local extension in Chrome
1. Hard navigate to https://www.youtube.com/watch?v=7a9lsGtVziM

If dynamic navigation is used for the last step, for example clicking on the video from another video, then the button will be injected as expected. Otherwise with static navigation, the button isn't being injected.

I'll be able to help with the debugging later this week.

**Comment by 1natsu172** [comment]
Well...apparently it seems that whether or not we are logged into YouTube affects the behavior. I use a separate Chrome account for development, and I haven't logged into YouTube over there. As I tried with YouTube logged in then static navigation to video, so didn't work as you said.

The specific cause is yet to be understood. Please help me if I can't solve it by the weekend…! 🙏

**Comment by dideler** [comment]
> Well...apparently it seems that whether or not we are logged into YouTube affects the behavior.

😱 

That was a tricky one to find, good spot.

I think it's time to introduce some automated tests (#9). There are a lot of cases we have to cover now and doing it manually is becoming too time consuming. I'll try starting that this weekend. 🤞

File: script.js [hunk 5]
```diff
@@ -175,7 +185,10 @@ const newYouTube = {
     return (
       typeof e !== 'undefined' &&
       e.type === 'yt-visibility-refresh' &&
-      e.target.tagName === 'YTD-ITEM-SECTION-RENDERER' // this mean: rendered childNode of comments'DOM(<ytd-comments>)
+      (
+        e.target.tagName === 'YTD-ITEM-SECTION-RENDERER' || // this mean: rendered childNode of comments'DOM(<ytd-comments>)
+        e.target.tagName === 'YTD-CREATOR-HEART-RENDERER' // this case: video type with heart action to user comments by creator.
+      )
```
Comment (1natsu172) [inline script.js:191]: Fix problem the user logged in YouTube and watching video with heart-action-comment page.

**Comment by 1natsu172** [comment]
@dideler, I fixed an issue where the logged in user and  the un-logged in user behaved differently.

My local is working well in either state, on the issue video page! Please check this again. :crossed_fingers:

**Comment by dideler** [comment]
Hey @1natsu172, sorry for the slow reply and thanks for your patience!

I tried the fix with the latest changes and it worked on video `7a9lsGtVziM` 🙌
But I was wondering if there was a simpler way without having to introduce timers.

I listened to every YT event I could find

image

```js
const events = [
  'yt-action',
  'yt-add-element-to-app',
  'yt-autonav-pause-blur',
  'yt-autonav-pause-focus',
  'yt-autonav-pause-guide-closed',
  'yt-autonav-pause-guide-open',
  'yt-autonav-pause-player',
  'yt-autonav-pause-player-ended',
  'yt-autonav-pause-player-scroll',
  'yt-focus-searchbox',
  'yt-guide-close',
  'yt-guide-toggle',
  'yt-history-load',
  'yt-history-pop',
  'yt-masthead-height-changed',
  'yt-navigate',
  'yt-navigate-finish',
  'yt-navigate-set-page-offset',
  'yt-navigate-start',
  'yt-page-type-changed',
  'yt-page-data-updated',
  'yt-preconnect-urls',
  'yt-register-action',
  'yt-report-form-closed',
  'yt-report-form-opened',
  'yt-request-panel-mode-change',
  'yt-retrieve-location',
  'yt-service-request-completed',
  'yt-set-theater-mode-enabled',
  'yt-swatch-changed',
  'yt-theater-mode-allowed',
  'yt-unregister-action',
  'yt-update-title',
  'yt-update-unseen-notification-count',
  'yt-visibility-refresh',
];

events.forEach(eventName => {
  console.log(`listening for ${eventName}`);
  document.addEventListener(eventName, e => {
    console.log(e);
  });
});
```

And I noticed that the event `yt-register-action` with tag `YTD-COMMENTS` was firing right before comments would be loaded.

See the result in #35 and give it a try. It also benefits from not showing the loading animation by the comments which would appear with the timer strategy, which could confuse users as it looks like comments are about to show and then they are hidden.

**Comment by 1natsu172** [comment]
Thanks for checking.
I don't think the timer strategy is a good solution. `yt-visibility-refresh` had no choice.
But I think your solution(#35) is perfect. Let my PR abolish and adopt your #35.

---

As an aside, I had seen a lot of "yt-*" events when I implemented the number of comments, but it seems I didn't have `yt-register-action` with tag `YTD-COMMENTS` before, or did I miss it? 🤔

**Comment by dideler** [comment]
> but it seems I didn't have `yt-register-action` with tag `YTD-COMMENTS` before, or did I miss it? 

Interestingly, the event `yt-register-action` was poorly being used by the extension to support New YT (ebb3290aa2788c759af10dd58433175e09173dc0) but was later removed in favour of `yt-visibility-refresh` (0dc8bfba2297574693907bc8826ecf2c91b6cfa8).

I'm not sure if `yt-register-action` always had `YTD-COMMENTS` as a tag, it could be something new.

**Pull request closed without merging**